framework/src/modules/chain/bft/consensus_manager.js

```diff
@@ -0,0 +1,297 @@
+
+		// Load or initialize delegate state in reference to current BlockHeaderManager block headers
+		const delegateState = this.state[delegatePublicKey] || {
+			maxPreVoteHeight: 0,
+			maxPreCommitHeight: 0,
+		};
+
+		// Get first block of the round when delegate was active
+		const delegateMinHeightActive =
+			(header.activeSinceRound - 1) * this.activeDelegates + 1;
+
+		// If delegate is new then first block of the round will be considered
+		// if it forged before then we probably have the last commit height
+		// delegate can't pre-commit and block before above mentioned conditions
+		const minPreCommit = Math.max(
+			delegateMinHeightActive,
+			delegateState.maxPreCommitHeight + 1
+		);
+
+		// Delegate can't pre-commit the blocks on tip of the chain
+		const maxPreCommitHeight = header.height - 1;
+
+		for (let j = minPreCommit; j <= maxPreCommitHeight; j++) {
+			// Add pre-commit if threshold is reached
+			if (this.preVotes[j] >= this.preVoteThreshold) {
+				// Increase the pre-commit for particular height
+				this.preCommits[j] = (this.preCommits[j] || 0) + 1;
+
+				// Keep track of the last pre-commit point
+				delegateState.maxPreCommitHeight = j;
+			}
+		}
+
+		// Check between height of first block of the round when delegate was active
+		// Or one step ahead where it forged the last block
+		// Or one step ahead where it left the last pre-vote
+		// Or maximum 3 rounds backward
+		const minPreVoteHeight = Math.max(
+			delegateMinHeightActive,
+			header.maxHeightPreviouslyForged + 1,
+			delegateState.maxPreVoteHeight + 1,
+			header.height - this.processingThreshold
+		);
+		// Pre-vote upto current block height
+		const maxPreVoteHeight = header.height;
+		for (let j = minPreVoteHeight; j <= maxPreVoteHeight; j++) {
+			this.preVotes[j] = (this.preVotes[j] || 0) + 1;
+		}
+		// Update delegate state
+		delegateState.maxPreVoteHeight = maxPreVoteHeight;
+
+		// Set the delegate state
+		this.state[delegatePublicKey] = delegateState;
+	}
+
+	/**
+	 * Update the pre-voted confirmed and finalized height
+	 */
+	updatePreVotedAndFinalizedHeight() {
+		debug('updatePreVotedAndFinalizedHeight invoked');
+		if (this.headers.length === 0) {
+			return;
+		}
+
+		const higherPairVoted = Object.entries(this.preVotes)
+			.reverse()
+			.find(pair => pair[1] >= this.preVoteThreshold);
+
+		this.prevotedConfirmedHeight = higherPairVoted
+			? parseInt(higherPairVoted[0])
+			: this.prevotedConfirmedHeight;
+
+		const higherPairCommitted = Object.entries(this.preCommits)
+			.reverse()
+			.find(pair => pair[1] >= this.preCommitThreshold);
+
+		this.finalizedHeight = higherPairCommitted
+			? parseInt(higherPairCommitted[0])
+			: this.finalizedHeight;
+	}
+
+	/**
+	 * Use existing block headers and re-compute all information
+	 */
+	recompute() {
+		this.state = {};
+		this.finalizedHeight = 0;
+		this.prevotedConfirmedHeight = 0;
+		this.preVotes = {};
+		this.preCommits = {};
+
+		this.headers.items.forEach(header => {
```

Comment:
Avoid using `forEach` when possible. I suggest using `for..of`

Reply:
Any particular reason for not using a default iterator function?

Reply:
`forEach` is proven to be the slowest way to iterate over an array, so in my opinion not using it by default is a good exercise.

Reply:
@limiaspasdaniel Created a separate issue to handle this and add more tests.